### PR TITLE
Deployment and Replica Set views

### LINKED ISSUE #1372: k8s: Add views for Deployments and ReplicaSets
Kubernetes 1.2 comes with [Deployments](http://kubernetes.io/docs/user-guide/deployments/) and [ReplicaSets](http://kubernetes.io/docs/user-guide/replicasets/) (similar to ReplicationControllers but enable declarative changes/rollouts and use hashing for versioning), we should add views for them.

See http://blog.kubernetes.io/2016/04/using-deployment-objects-with.html for a friendly example.

For Deployments we probably want some sort of log view to see the changes which were made to the deployment (it would require a similar view to https://github.com/weaveworks/scope/issues/1329 )

### PR DESCRIPTION
**Pull request by paulbellamy**
Fixes #1372

### CODE REVIEW

File: probe/kubernetes/meta.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/labels"
+)
+
+// Meta represents a metadata information about a Kubernetes object
+type Meta interface {
```
Comment (tomwilkie) [inline probe/kubernetes/meta.go]: Nice!

**Comment by tomwilkie** [comment]
I think we should either include a replication controller view, or include replication controllers in replica sets.

People still have replication controllers.  Like us.

File: probe/kubernetes/deployment.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	"github.com/weaveworks/scope/report"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/labels"
+)
+
+// These constants are keys used in node metadata
+const (
+	DeploymentID                  = "kubernetes_deployment_id"
```
Comment (tomwilkie) [inline probe/kubernetes/deployment.go]: Given we don't merge nodes of different types anymore, is it worth have a meta.GetNode() and unifying all the metadata keys between deployments, pods, services, rs, rc etc etc?

File: render/pod.go [hunk 3]
```diff
@@ -57,6 +57,34 @@ var PodServiceRenderer = ConditionalRenderer(renderKubernetesTopologies,
...
+var DeploymentRenderer = ApplyDecorators(
+	FilterEmpty(report.Pod,
+		MakeReduce(
+			MakeMap(
+				MapPod2Deployment,
+				PodRenderer,
+			),
+			SelectDeployment,
+		),
+	),
```
Comment (tomwilkie) [inline render/pod.go]: Is this right?  I was under the impression is was pod -> replica set -> deployment?

**Comment by tomwilkie** [comment]
Looks good so far!  I'll work on controls for this tomorrow.

**Comment by paulbellamy** [comment]
Also, need to check this with multiple nodes, as the deployments/replicasets/services will get reported multiple times

**Comment by tomwilkie** [comment]
GKE ftw

File: render/pod.go (hunk 3 above)
Comment (paulbellamy) [inline render/pod.go]: The selectors on a deployment match either. Technically a deployment manages a replicaset which manages the pods, but in practice deployments and replica sets are somewhat interchangeable (with the difference that they tell you not to use replica sets). However, showing the full hierarchy may be more clear.

**Comment by tomwilkie** [comment]
@paulbellamy I just deployed this on GKE.

It looks really good.

File: probe/kubernetes/meta.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+func (m meta) Created() string {
+	return m.ObjectMeta.CreationTimestamp.Format(time.RFC822)
+}
+
+func (m meta) Labels() map[string]string {
+	return m.ObjectMeta.Labels
+}
+
+// MetaNode gets the node metadata
+func MetaNode(m Meta, id string) report.Node {
```
Comment (tomwilkie) [inline probe/kubernetes/meta.go]: nit: why not:

```
func (m Meta) MetaNode(id string) report.Node
```

**Comment by tomwilkie** [comment]
Other than the nit, LGTM

**Pull request merged**